File: apps/quant-web/src/utils/rangeDetectorLux.ts
import type { BarData } from '@/types/market'

export type RangeDetectorState = 'intact' | 'broken_up' | 'broken_down'
export type RangeDetectorTransitionKind = 'confirmed' | 'revised' | 'broken_up' | 'broken_down' | 'invalid_reset'
export interface RangeDetectorLuxOptions { sourceIdentity: string; minimumRangeLength?: number; rangeWidthAtrMultiplier?: number; rangeAtrLength?: number; roundDigits?: number }
export interface RangeDetectorSnapshot { formulaVersion: 'range_detector_lux_v1'; policyId: 'range_detector_lux_v1'; rangeId: string; revision: number; visualStartAt: string; confirmedAt: string; detectionRightAt: string; levelsActiveFrom: string; initialUpper: number; initialLower: number; currentUpper: number; currentLower: number; currentMid: number; state: RangeDetectorState; brokenAt: string | null; mergedCount: number; candidateValid: boolean; sourceBarEnd: string; sourceTradingDay: string | null; sourceIdentity: string }
export interface RangeDetectorPoint { time: string; ready: boolean; valid: boolean; reason: string | null; snapshot: RangeDetectorSnapshot | null; transition: { kind: RangeDetectorTransitionKind; rangeId: string | null; revision: number | null; at: string } | null }
export interface RangeDetectorVisualRange { key: string; rangeId: string; revision: number; visualStartAt: string; detectionRightAt: string; levelsActiveFrom: string; levelsActiveUntil: string | null; confirmedAt: string; upper: number; lower: number; mid: number; state: RangeDetectorState; brokenAt: string | null }
export interface RangeDetectorLuxResult { points: RangeDetectorPoint[]; ranges: RangeDetectorVisualRange[] }

interface Atr { seed: number[]; previousClose: number | null; previous: number | null }
interface Active { snapshot: RangeDetectorSnapshot | null; detectionRightIndex: number | null }

export function calculateRangeDetectorLux(bars: BarData[], options: RangeDetectorLuxOptions): RangeDetectorLuxResult {
  const minimumRangeLength = options.minimumRangeLength ?? 20
  const rangeWidthAtrMultiplier = options.rangeWidthAtrMultiplier ?? 1
  const rangeAtrLength = options.rangeAtrLength ?? 500
  const roundDigits = options.roundDigits ?? 6
  if (!options.sourceIdentity?.trim()) throw new Error('sourceIdentity must be non-empty')
  if (!Number.isInteger(minimumRangeLength) || minimumRangeLength < 2) throw new Error('minimumRangeLength must be at least 2')
  if (!Number.isInteger(rangeAtrLength) || rangeAtrLength <= 0) throw new Error('rangeAtrLength must be positive')
  if (!Number.isFinite(rangeWidthAtrMultiplier) || rangeWidthAtrMultiplier <= 0) throw new Error('rangeWidthAtrMultiplier must be finite and positive')
  if (!Number.isInteger(roundDigits) || roundDigits < 0) throw new Error('roundDigits must be non-negative')

  const points: RangeDetectorPoint[] = []
  let atr: Atr = { seed: [], previousClose: null, previous: null }
  let window: Array<{ time: string; tradingDay: string | null; close: number }> = []
  let previousCandidate = false
  let active: Active = { snapshot: null, detectionRightIndex: null }
  let lastTime: number | null = null
  for (let index = 0; index < bars.length; index += 1) {
    const bar = bars[index]
    const timestamp = parseTime(bar.time)
    if (lastTime !== null && timestamp <= lastTime) throw new Error('bar times must be strictly increasing')
    lastTime = timestamp
    const valid = Number.isFinite(bar.high) && Number.isFinite(bar.low) && Number.isFinite(bar.close) && bar.low <= bar.close && bar.close <= bar.high
    if (!valid) {
      atr = { seed: [], previousClose: null, previous: null }; window = []; previousCandidate = false; active = { snapshot: null, detectionRightIndex: null }
      points.push({ time: bar.time, ready: true, valid: false, reason: 'input_invalid', snapshot: null, transition: { kind: 'invalid_reset', rangeId: null, revision: null, at: bar.time } })
      continue
    }
    const tr = atr.previousClose === null ? bar.high - bar.low : Math.max(bar.high - bar.low, Math.abs(bar.high - atr.previousClose), Math.abs(bar.low - atr.previousClose))
    atr.seed = [...atr.seed, tr].slice(-rangeAtrLength); atr.previousClose = bar.close
    if (atr.previous === null) { if (atr.seed.length === rangeAtrLength) atr.previous = average(atr.seed) } else atr.previous = (atr.previous * (rangeAtrLength - 1) + tr) / rangeAtrLength
    window = [...window, { time: bar.time, tradingDay: bar.trading_day ?? null, close: bar.close }].slice(-(minimumRangeLength + 1))
    if (atr.previous === null || window.length !== minimumRangeLength + 1) {
      previousCandidate = false
      points.push({ time: bar.time, ready: false, valid: true, reason: 'warming_up', snapshot: null, transition: null })
      continue
    }
    const candidateWindow = window.slice(-minimumRangeLength); const center = average(candidateWindow.map((item) => item.close)); const width = atr.previous * rangeWidthAtrMultiplier
    const candidate = candidateWindow.every((item) => Math.abs(item.close - center) <= width)
    let snapshot = active.snapshot; let transition: RangeDetectorPoint['transition'] = null
    if (candidate && !previousCandidate) {
      const upper = rounded(center + width, roundDigits); const lower = rounded(center - width, roundDigits)
      const overlap = snapshot !== null && active.detectionRightIndex !== null && index - minimumRangeLength <= active.detectionRightIndex
      if (overlap && snapshot) {
        const currentUpper = Math.max(snapshot.currentUpper, upper); const currentLower = Math.min(snapshot.currentLower, lower)
        snapshot = { ...snapshot, revision: snapshot.revision + 1, confirmedAt: bar.time, detectionRightAt: bar.time, levelsActiveFrom: bar.time, currentUpper, currentLower, currentMid: rounded((currentUpper + currentLower) / 2, roundDigits), state: 'intact', brokenAt: null, mergedCount: snapshot.mergedCount + 1, candidateValid: true, sourceBarEnd: bar.time, sourceTradingDay: bar.trading_day ?? null }
        transition = { kind: 'revised', rangeId: snapshot.rangeId, revision: snapshot.revision, at: bar.time }
      } else {
        const rangeId = sha256(`range_detector_lux_v1|${options.sourceIdentity}|${bar.time}`)
        snapshot = { formulaVersion: 'range_detector_lux_v1', policyId: 'range_detector_lux_v1', rangeId, revision: 1, visualStartAt: window[0].time, confirmedAt: bar.time, detectionRightAt: bar.time, levelsActiveFrom: bar.time, initialUpper: upper, initialLower: lower, currentUpper: upper, currentLower: lower, currentMid: rounded(center, roundDigits), state: 'intact', brokenAt: null, mergedCount: 0, candidateValid: true, sourceBarEnd: bar.time, sourceTradingDay: bar.trading_day ?? null, sourceIdentity: options.sourceIdentity }
        transition = { kind: 'confirmed', rangeId, revision: 1, at: bar.time }
      }
      active.detectionRightIndex = index
    } else if (candidate && snapshot) { snapshot = { ...snapshot, detectionRightAt: bar.time, candidateValid: true, sourceBarEnd: bar.time, sourceTradingDay: bar.trading_day ?? null }; active.detectionRightIndex = index } else if (snapshot) snapshot = { ...snapshot, candidateValid: false, sourceBarEnd: bar.time, sourceTradingDay: bar.trading_day ?? null }
    if (snapshot?.state === 'intact' && bar.close > snapshot.currentUpper) { snapshot = { ...snapshot, state: 'broken_up', brokenAt: bar.time }; transition = { kind: 'broken_up', rangeId: snapshot.rangeId, revision: snapshot.revision, at: bar.time } }
    else if (snapshot?.state === 'intact' && bar.close < snapshot.currentLower) { snapshot = { ...snapshot, state: 'broken_down', brokenAt: bar.time }; transition = { kind: 'broken_down', rangeId: snapshot.rangeId, revision: snapshot.revision, at: bar.time } }
    active.snapshot = snapshot; previousCandidate = candidate
    points.push({ time: bar.time, ready: true, valid: true, reason: null, snapshot, transition })
  }
  return { points, ranges: visualRanges(points) }
}

function visualRanges(points: RangeDetectorPoint[]): RangeDetectorVisualRange[] {
  const snapshots = new Map<string, RangeDetectorSnapshot>(); const order: string[] = []; const terminal = new Map<string, string>(); let latest: string | null = null
  for (const point of points) {
    if (point.transition?.kind === 'invalid_reset') { if (latest) terminal.set(latest, point.time); latest = null }
    const snapshot = point.snapshot; if (!snapshot) continue; const key = `${snapshot.rangeId}:${snapshot.revision}`
    if (!snapshots.has(key)) { if (latest && latest !== key) terminal.set(latest, snapshot.levelsActiveFrom); order.push(key) }
    snapshots.set(key, snapshot); latest = key
  }
  return order.map((key) => { const item = snapshots.get(key)!; return { key, rangeId: item.rangeId, revision: item.revision, visualStartAt: item.visualStartAt, detectionRightAt: item.detectionRightAt, levelsActiveFrom: item.levelsActiveFrom, levelsActiveUntil: terminal.get(key) ?? null, confirmedAt: item.confirmedAt, upper: item.currentUpper, lower: item.currentLower, mid: item.currentMid, state: item.state, brokenAt: item.brokenAt } })
}
function parseTime(value: string): number { if (!value.includes('T') || Number.isNaN(Date.parse(value))) throw new Error('bar time must be ISO-8601'); return Date.parse(value) }
function average(values: number[]): number { return values.reduce((sum, value) => sum + value, 0) / values.length }
function rounded(value: number, digits: number): number { return Number(value.toFixed(digits)) }

function sha256(input: string): string {
  const bytes = Array.from(new TextEncoder().encode(input)); const bitLength = bytes.length * 8; bytes.push(0x80); while ((bytes.length % 64) !== 56) bytes.push(0); for (let shift = 56; shift >= 0; shift -= 8) bytes.push((bitLength / 2 ** shift) & 255)
  const hash = [1779033703, 3144134277, 1013904242, 2773480762, 1359893119, 2600822924, 528734635, 1541459225]; const constants = [1116352408,1899447441,3049323471,3921009573,961987163,1508970993,2453635748,2870763221,3624381080,310598401,607225278,1426881987,1925078388,2162078206,2614888103,3248222580,3835390401,4022224774,264347078,604807628,770255983,1249150122,1555081692,1996064986,2554220882,2821834349,2952996808,3210313671,3336571891,3584528711,113926993,338241895,666307205,773529912,1294757372,1396182291,1695183700,1986661051,2177026350,2456956037,2730485921,2820302411,3259730800,3345764771,3516065817,3600352804,4094571909,275423344,430227734,506948616,659060556,883997877,958139571,1322822218,1537002063,1747873779,1955562222,2024104815,2227730452,2361852424,2428436474,2756734187,3204031479,3329325298]
  for (let offset = 0; offset < bytes.length; offset += 64) { const words = Array(64).fill(0); for (let index = 0; index < 16; index += 1) words[index] = (bytes[offset + index * 4] << 24) | (bytes[offset + index * 4 + 1] << 16) | (bytes[offset + index * 4 + 2] << 8) | bytes[offset + index * 4 + 3]; for (let index = 16; index < 64; index += 1) { const a = words[index - 15]; const b = words[index - 2]; words[index] = (((a >>> 7 | a << 25) ^ (a >>> 18 | a << 14) ^ a >>> 3) + words[index - 7] + ((b >>> 17 | b << 15) ^ (b >>> 19 | b << 13) ^ b >>> 10) + words[index - 16]) | 0 } let [a,b,c,d,e,f,g,h] = hash; for (let index = 0; index < 64; index += 1) { const s1 = (e >>> 6 | e << 26) ^ (e >>> 11 | e << 21) ^ (e >>> 25 | e << 7); const choice = (e & f) ^ (~e & g); const temp1 = (h + s1 + choice + constants[index] + words[index]) | 0; const s0 = (a >>> 2 | a << 30) ^ (a >>> 13 | a << 19) ^ (a >>> 22 | a << 10); const majority = (a & b) ^ (a & c) ^ (b & c); [h,g,f,e,d,c,b,a] = [g,f,e,(d + temp1) | 0,c,b,a,(temp1 + s0 + majority) | 0] } hash[0]=(hash[0]+a)|0; hash[1]=(hash[1]+b)|0; hash[2]=(hash[2]+c)|0; hash[3]=(hash[3]+d)|0; hash[4]=(hash[4]+e)|0; hash[5]=(hash[5]+f)|0; hash[6]=(hash[6]+g)|0; hash[7]=(hash[7]+h)|0 }
  return hash.map((value) => (value >>> 0).toString(16).padStart(8, '0')).join('')
}

File: packages/quant-core/guiyi_quant/indicators/range_detector_lux.py
from __future__ import annotations

import hashlib
import math
from collections.abc import Sequence
from dataclasses import dataclass, replace
from datetime import datetime
from typing import Literal

from .atr import initial_atr_state, step_atr
from .models import AtrState


RANGE_DETECTOR_LUX_CODE = "range_detector_lux_v1"
RANGE_DETECTOR_LUX_VERSION = "v1"
RANGE_DETECTOR_LUX_POLICY_ID = "range_detector_lux_v1"

RangeDetectorState = Literal["intact", "broken_up", "broken_down"]
RangeDetectorTransitionKind = Literal[
    "confirmed",
    "revised",
    "broken_up",
    "broken_down",
    "invalid_reset",
]


@dataclass(frozen=True, slots=True)
class RangeDetectorLuxParameters:
    minimum_range_length: int = 20
    range_width_atr_multiplier: float = 1.0
    range_atr_length: int = 500
    round_digits: int = 6


@dataclass(frozen=True, slots=True)
class RangeDetectorSnapshot:
    formula_version: str
    policy_id: str
    range_id: str
    revision: int
    visual_start_at: str
    confirmed_at: str
    detection_right_at: str
    levels_active_from: str
    initial_upper: float
    initial_lower: float
    current_upper: float
    current_lower: float
    current_mid: float
    state: RangeDetectorState
    broken_at: str | None
    merged_count: int
    candidate_valid: bool
    source_bar_end: str
    source_trading_day: str | None
    source_identity: str


@dataclass(frozen=True, slots=True)
class RangeDetectorVisualRange:
    range_id: str
    revision: int
    visual_start_at: str
    detection_right_at: str
    levels_active_from: str
    levels_active_until: str | None
    confirmed_at: str
    upper: float
    lower: float
    mid: float
    state: RangeDetectorState
    broken_at: str | None


@dataclass(frozen=True, slots=True)
class RangeDetectorTransition:
    kind: RangeDetectorTransitionKind
    range_id: str | None
    revision: int | None
    at: str


@dataclass(frozen=True, slots=True)
class RangeDetectorPoint:
    bar_end: str
    ready: bool
    valid: bool
    reason: str | None
    snapshot: RangeDetectorSnapshot | None
    transition: RangeDetectorTransition | None


@dataclass(frozen=True, slots=True)
class RangeDetectorLuxState:
    parameters: RangeDetectorLuxParameters
    source_identity: str
    atr: AtrState
    index: int
    close_window: tuple[tuple[str, str | None, float], ...]
    previous_candidate_valid: bool
    active_snapshot: RangeDetectorSnapshot | None
    active_detection_right_index: int | None
    last_bar_end: str | None


@dataclass(frozen=True, slots=True)
class RangeDetectorSeries:
    indicator_code: str
    indicator_version: str
    policy_id: str
    parameters: RangeDetectorLuxParameters
    source_identity: str
    points: tuple[RangeDetectorPoint, ...]
    ranges: tuple[RangeDetectorVisualRange, ...]


def initial_range_detector_lux_state(
    *,
    source_identity: str,
    minimum_range_length: int = 20,
    range_width_atr_multiplier: float = 1.0,
    range_atr_length: int = 500,
    round_digits: int = 6,
) -> RangeDetectorLuxState:
    parameters = RangeDetectorLuxParameters(
        minimum_range_length=minimum_range_length,
        range_width_atr_multiplier=range_width_atr_multiplier,
        range_atr_length=range_atr_length,
        round_digits=round_digits,
    )
    _validate_parameters(parameters, source_identity)
    return RangeDetectorLuxState(
        parameters=parameters,
        source_identity=source_identity,
        atr=initial_atr_state(
            range_atr_length,
            smoothing_policy="wilder_sma_seed",
            round_digits=round_digits,
        ),
        index=-1,
        close_window=(),
        previous_candidate_valid=False,
        active_snapshot=None,
        active_detection_right_index=None,
        last_bar_end=None,
    )


def step_range_detector_lux(
    state: RangeDetectorLuxState,
    *,
    high: float | int | None,
    low: float | int | None,
    close: float | int | None,
    bar_end: str,
    trading_day: str | None = None,
) -> tuple[RangeDetectorLuxState, RangeDetectorPoint]:
    """Advance a causal Range Detector Lux V1 state by one completed bar."""

    timestamp = _parse_bar_end(bar_end)
    if state.last_bar_end is not None and timestamp <= _parse_bar_end(state.last_bar_end):
        raise ValueError("bar_end values must be strictly increasing")

    next_index = state.index + 1
    high_value = _finite_float(high)
    low_value = _finite_float(low)
    close_value = _finite_float(close)
    input_valid = (
        high_value is not None
        and low_value is not None
        and close_value is not None
        and low_value <= close_value <= high_value
    )
    atr, atr_point = step_atr(
        state.atr,
        high=high_value if input_valid else None,
        low=low_value if input_valid else None,
        close=close_value if input_valid else None,
        bar_end=bar_end,
    )
    if not input_valid:
        invalid_transition = RangeDetectorTransition(
            kind="invalid_reset",
            range_id=None,
            revision=None,
            at=bar_end,
        )
        return (
            replace(
                state,
                atr=atr,
                index=next_index,
                close_window=(),
                previous_candidate_valid=False,
                active_snapshot=None,
                active_detection_right_index=None,
                last_bar_end=bar_end,
            ),
            RangeDetectorPoint(
                bar_end=bar_end,
                ready=True,
                valid=False,
                reason="input_invalid",
                snapshot=None,
                transition=invalid_transition,
            ),
        )

    assert close_value is not None
    close_window = (
        *state.close_window,
        (bar_end, trading_day, close_value),
    )[-(state.parameters.minimum_range_length + 1) :]
    ready = atr_point.ready and len(close_window) == state.parameters.minimum_range_length + 1
    if not ready:
        return (
            replace(
                state,
                atr=atr,
                index=next_index,
                close_window=close_window,
                previous_candidate_valid=False,
                last_bar_end=bar_end,
            ),
            RangeDetectorPoint(
                bar_end=bar_end,
                ready=False,
                valid=True,
                reason="warming_up",
                snapshot=None,
                transition=None,
            ),
        )

    assert atr.previous_atr is not None
    candidate_window = close_window[-state.parameters.minimum_range_length :]
    center = sum(value for _, _, value in candidate_window) / len(candidate_window)
    width = atr.previous_atr * state.parameters.range_width_atr_multiplier
    candidate_valid = all(
        abs(value - center) <= width for _, _, value in candidate_window
    )
    snapshot = state.active_snapshot
    active_detection_right_index = state.active_detection_right_index
    transition: RangeDetectorTransition | None = None

    if candidate_valid and not state.previous_candidate_valid:
        candidate_upper = _round(center + width, state.parameters.round_digits)
        candidate_lower = _round(center - width, state.parameters.round_digits)
        visual_start_at = close_window[0][0]
        visual_start_index = next_index - state.parameters.minimum_range_length
        if (
            snapshot is not None
            and active_detection_right_index is not None
            and visual_start_index <= active_detection_right_index
        ):
            current_upper = max(snapshot.current_upper, candidate_upper)
            current_lower = min(snapshot.current_lower, candidate_lower)
            snapshot = RangeDetectorSnapshot(
                formula_version=RANGE_DETECTOR_LUX_CODE,
                policy_id=RANGE_DETECTOR_LUX_POLICY_ID,
                range_id=snapshot.range_id,
                revision=snapshot.revision + 1,
                visual_start_at=snapshot.visual_start_at,
                confirmed_at=bar_end,
                detection_right_at=bar_end,
                levels_active_from=bar_end,
                initial_upper=snapshot.initial_upper,
                initial_lower=snapshot.initial_lower,
                current_upper=current_upper,
                current_lower=current_lower,
                current_mid=_round(
                    (current_upper + current_lower) / 2,
                    state.parameters.round_digits,
                ),
                state="intact",
                broken_at=None,
                merged_count=snapshot.merged_count + 1,
                candidate_valid=True,
                source_bar_end=bar_end,
                source_trading_day=trading_day,
                source_identity=state.source_identity,
            )
            transition = RangeDetectorTransition(
                kind="revised",
                range_id=snapshot.range_id,
                revision=snapshot.revision,
                at=bar_end,
            )
        else:
            range_id = _range_id(state.source_identity, bar_end)
            snapshot = RangeDetectorSnapshot(
                formula_version=RANGE_DETECTOR_LUX_CODE,
                policy_id=RANGE_DETECTOR_LUX_POLICY_ID,
                range_id=range_id,
                revision=1,
                visual_start_at=visual_start_at,
                confirmed_at=bar_end,
                detection_right_at=bar_end,
                levels_active_from=bar_end,
                initial_upper=candidate_upper,
                initial_lower=candidate_lower,
                current_upper=candidate_upper,
                current_lower=candidate_lower,
                current_mid=_round(center, state.parameters.round_digits),
                state="intact",
                broken_at=None,
                merged_count=0,
                candidate_valid=True,
                source_bar_end=bar_end,
                source_trading_day=trading_day,
                source_identity=state.source_identity,
            )
            transition = RangeDetectorTransition(
                kind="confirmed",
                range_id=range_id,
                revision=1,
                at=bar_end,
            )
        active_detection_right_index = next_index
    elif candidate_valid and snapshot is not None:
        snapshot = replace(
            snapshot,
            detection_right_at=bar_end,
            candidate_valid=True,
            source_bar_end=bar_end,
            source_trading_day=trading_day,
        )
        active_detection_right_index = next_index
    elif snapshot is not None:
        snapshot = replace(
            snapshot,
            candidate_valid=False,
            source_bar_end=bar_end,
            source_trading_day=trading_day,
        )

    if snapshot is not None and snapshot.state == "intact":
        if close_value > snapshot.current_upper:
            snapshot = replace(snapshot, state="broken_up", broken_at=bar_end)
            transition = RangeDetectorTransition(
                kind="broken_up",
                range_id=snapshot.range_id,
                revision=snapshot.revision,
                at=bar_end,
            )
        elif close_value < snapshot.current_lower:
            snapshot = replace(snapshot, state="broken_down", broken_at=bar_end)
            transition = RangeDetectorTransition(
                kind="broken_down",
                range_id=snapshot.range_id,
                revision=snapshot.revision,
                at=bar_end,
            )

    return (
        replace(
            state,
            atr=atr,
            index=next_index,
            close_window=close_window,
            previous_candidate_valid=candidate_valid,
            active_snapshot=snapshot,
            active_detection_right_index=active_detection_right_index,
            last_bar_end=bar_end,
        ),
        RangeDetectorPoint(
            bar_end=bar_end,
            ready=True,
            valid=True,
            reason=None,
            snapshot=snapshot,
            transition=transition,
        ),
    )


def range_detector_lux_series(
    highs: Sequence[float | int | None],
    lows: Sequence[float | int | None],
    closes: Sequence[float | int | None],
    *,
    bar_ends: Sequence[str],
    source_identity: str,
    trading_days: Sequence[str | None] | None = None,
    minimum_range_length: int = 20,
    range_width_atr_multiplier: float = 1.0,
    range_atr_length: int = 500,
    round_digits: int = 6,
) -> RangeDetectorSeries:
    if len(highs) != len(lows) or len(highs) != len(closes):
        raise ValueError("highs, lows, and closes length must match")
    if len(bar_ends) != len(highs):
        raise ValueError("bar_ends length must match highs length")
    if trading_days is not None and len(trading_days) != len(highs):
        raise ValueError("trading_days length must match highs length")

    state = initial_range_detector_lux_state(
        source_identity=source_identity,
        minimum_range_length=minimum_range_length,
        range_width_atr_multiplier=range_width_atr_multiplier,
        range_atr_length=range_atr_length,
        round_digits=round_digits,
    )
    points: list[RangeDetectorPoint] = []
    for index, (high, low, close, bar_end) in enumerate(
        zip(highs, lows, closes, bar_ends, strict=True)
    ):
        state, point = step_range_detector_lux(
            state,
            high=high,
            low=low,
            close=close,
            bar_end=bar_end,
            trading_day=None if trading_days is None else trading_days[index],
        )
        points.append(point)

    return RangeDetectorSeries(
        indicator_code=RANGE_DETECTOR_LUX_CODE,
        indicator_version=RANGE_DETECTOR_LUX_VERSION,
        policy_id=RANGE_DETECTOR_LUX_POLICY_ID,
        parameters=state.parameters,
        source_identity=source_identity,
        points=tuple(points),
        ranges=_visual_ranges(points),
    )


def _visual_ranges(
    points: Sequence[RangeDetectorPoint],
) -> tuple[RangeDetectorVisualRange, ...]:
    snapshots: dict[tuple[str, int], RangeDetectorSnapshot] = {}
    order: list[tuple[str, int]] = []
    terminal_at: dict[tuple[str, int], str] = {}
    latest_key: tuple[str, int] | None = None

    for point in points:
        if point.transition is not None and point.transition.kind == "invalid_reset":
            if latest_key is not None:
                terminal_at.setdefault(latest_key, point.bar_end)
            latest_key = None
        snapshot = point.snapshot
        if snapshot is None:
            continue
        key = (snapshot.range_id, snapshot.revision)
        if key not in snapshots:
            if latest_key is not None and latest_key != key:
                terminal_at.setdefault(latest_key, snapshot.levels_active_from)
            order.append(key)
        snapshots[key] = snapshot
        latest_key = key

    visual_ranges: list[RangeDetectorVisualRange] = []
    for key in order:
        snapshot = snapshots[key]
        visual_ranges.append(
            RangeDetectorVisualRange(
                range_id=snapshot.range_id,
                revision=snapshot.revision,
                visual_start_at=snapshot.visual_start_at,
                detection_right_at=snapshot.detection_right_at,
                levels_active_from=snapshot.levels_active_from,
                levels_active_until=terminal_at.get(key),
                confirmed_at=snapshot.confirmed_at,
                upper=snapshot.current_upper,
                lower=snapshot.current_lower,
                mid=snapshot.current_mid,
                state=snapshot.state,
                broken_at=snapshot.broken_at,
            )
        )
    return tuple(visual_ranges)


def _validate_parameters(
    parameters: RangeDetectorLuxParameters,
    source_identity: str,
) -> None:
    if not isinstance(source_identity, str) or not source_identity.strip():
        raise ValueError("source_identity must be a non-empty string")
    if parameters.minimum_range_length < 2:
        raise ValueError("minimum_range_length must be at least 2")
    if parameters.range_atr_length <= 0:
        raise ValueError("range_atr_length must be positive")
    if (
        not math.isfinite(parameters.range_width_atr_multiplier)
        or parameters.range_width_atr_multiplier <= 0
    ):
        raise ValueError("range_width_atr_multiplier must be finite and positive")
    if parameters.round_digits < 0:
        raise ValueError("round_digits must be non-negative")


def _parse_bar_end(bar_end: str) -> datetime:
    if not isinstance(bar_end, str) or "T" not in bar_end:
        raise ValueError("bar_end must be ISO-8601")
    try:
        parsed = datetime.fromisoformat(
            f"{bar_end[:-1]}+00:00" if bar_end.endswith("Z") else bar_end
        )
    except ValueError as exc:
        raise ValueError("bar_end must be ISO-8601") from exc
    if parsed.tzinfo is None:
        raise ValueError("bar_end must be ISO-8601 with timezone")
    return parsed


def _finite_float(value: float | int | None) -> float | None:
    if value is None:
        return None
    number = float(value)
    return number if math.isfinite(number) else None


def _round(value: float, round_digits: int) -> float:
    if not math.isfinite(value):
        raise ValueError("rounding value must be finite")
    coefficient, decimal_exponent = _canonical_decimal_components(abs(value))
    target_exponent = -round_digits
    if decimal_exponent >= target_exponent:
        rounded_coefficient = coefficient * 10 ** (
            decimal_exponent - target_exponent
        )
    else:
        divisor = 10 ** (target_exponent - decimal_exponent)
        rounded_coefficient, remainder = divmod(coefficient, divisor)
        halfway = divisor // 2
        if remainder > halfway or (
            remainder == halfway and rounded_coefficient % 2 == 1
        ):
            rounded_coefficient += 1
    if value < 0:
        rounded_coefficient = -rounded_coefficient
    rounded = float(f"{rounded_coefficient}e{-round_digits}")
    return 0.0 if rounded == 0 else rounded


def _canonical_decimal_components(value: float) -> tuple[int, int]:
    mantissa, separator, raw_exponent = str(value).lower().partition("e")
    exponent = int(raw_exponent) if separator else 0
    whole, dot, fraction = mantissa.partition(".")
    digits = f"{whole}{fraction}" if dot else whole
    return int(digits), exponent - len(fraction)


def _range_id(source_identity: str, first_confirmed_at: str) -> str:
    payload = "|".join(
        (RANGE_DETECTOR_LUX_CODE, source_identity, first_confirmed_at)
    )
    return hashlib.sha256(payload.encode("utf-8")).hexdigest()
